sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/nodes/intrinsics/multithreading/LLVMPThreadKeyIntrinsics.java

```diff
@@ -0,0 +1,118 @@
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
+ * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.oracle.truffle.llvm.runtime.nodes.intrinsics.multithreading;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.dsl.CachedContext;
+import com.oracle.truffle.api.dsl.NodeChild;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.llvm.runtime.LLVMContext;
+import com.oracle.truffle.llvm.runtime.LLVMLanguage;
+import com.oracle.truffle.llvm.runtime.interop.access.LLVMInteropType;
+import com.oracle.truffle.llvm.runtime.nodes.api.LLVMExpressionNode;
+import com.oracle.truffle.llvm.runtime.nodes.api.LLVMStoreNode;
+import com.oracle.truffle.llvm.runtime.nodes.intrinsics.llvm.LLVMBuiltin;
+import com.oracle.truffle.llvm.runtime.pointer.LLVMNativePointer;
+import com.oracle.truffle.llvm.runtime.pointer.LLVMPointer;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+public final class LLVMPThreadKeyIntrinsics {
+
+    @NodeChild(type = LLVMExpressionNode.class, value = "key")
+    @NodeChild(type = LLVMExpressionNode.class, value = "destructor")
+    public abstract static class LLVMPThreadKeyCreate extends LLVMBuiltin {
+
+        @Child LLVMStoreNode store = null;
+
+        @Specialization
+        protected int doIntrinsic(LLVMPointer key, LLVMPointer destructor, @CachedContext(LLVMLanguage.class) LLVMContext ctx) {
+            if (store == null) {
+                CompilerDirectives.transferToInterpreterAndInvalidate();
+                store = ctx.getLanguage().getNodeFactory().createStoreNode(LLVMInteropType.ValueKind.I32);
+            }
+
+            synchronized (ctx.keyLockObj) {
+                store.executeWithTarget(key, ctx.curKeyVal + 1);
+                // add new key-value to key-storage, which is a
+                // hashmap(key-value->hashmap(thread-id->specific-value))
+                UtilAccessCollectionWithBoundary.put(ctx.keyStorage, ctx.curKeyVal + 1, new ConcurrentHashMap<>());
+                UtilAccessCollectionWithBoundary.put(ctx.destructorStorage, ctx.curKeyVal + 1, destructor);
+                ctx.curKeyVal++;
+            }
+            return 0;
+        }
+    }
+
+    @NodeChild(type = LLVMExpressionNode.class, value = "key")
+    public abstract static class LLVMPThreadKeyDelete extends LLVMBuiltin {
+
+        @Specialization
+        protected int doIntrinsic(int key, @CachedContext(LLVMLanguage.class) LLVMContext ctx) {
```

Comment:
Sulong typically uses `context` as name of a context variable. In the interest of consistency I would ask that you also use that. I also notices that you sometimes use throwaway names like `th` for variable names. To make your code better understandable please use more speaking variable names.

Reply:
ok, changed that and also the th, t1, t2 stuff